File: yurt-convention-api/src/main/java/com/github/liulus/yurt/convention/util/SpelUtils.java
package com.github.liulus.yurt.convention.util;

import org.springframework.context.expression.MapAccessor;
import org.springframework.expression.spel.standard.SpelExpressionParser;
import org.springframework.expression.spel.support.StandardEvaluationContext;

import java.util.Map;

/**
 * @author liulu
 * @version V1.0
 * @since 2020/11/9
 */
public abstract class SpelUtils {

    private static final SpelExpressionParser EXPRESSION_PARSER = new SpelExpressionParser();

    private static final MapAccessor MAP_ACCESSOR = new MapAccessor();

    private SpelUtils() {
    }

    public static Object getValue(String spel, Object root) {
        Asserts.notNull(root, "root object can not be null");
        StandardEvaluationContext context = new StandardEvaluationContext(root);
        if (root instanceof Map) {
            context.addPropertyAccessor(MAP_ACCESSOR);
        }
        return EXPRESSION_PARSER.parseExpression(spel).getValue(context);
    }

    public static <T> T getValue(String spel, Object root, Class<T> returnType) {
        Asserts.notNull(root, "root object can not be null");
        StandardEvaluationContext context = new StandardEvaluationContext(root);
        if (root instanceof Map) {
            context.addPropertyAccessor(MAP_ACCESSOR);
        }
        return EXPRESSION_PARSER.parseExpression(spel).getValue(context, returnType);
    }

}

File: yurt-convention-api/src/main/kotlin/com/github/liulus/yurt/convention/util/SpelUtils.kt
package com.github.liulus.yurt.convention.util

import org.springframework.context.expression.MapAccessor
import org.springframework.expression.spel.standard.SpelExpressionParser
import org.springframework.expression.spel.support.StandardEvaluationContext

/**
 *
 * @author liulu
 * @version V1.0
 * @since 2020/11/5
 */
object SpelUtils {

    @JvmStatic
    private val expressionParser = SpelExpressionParser()

    @JvmStatic
    private val mapAccessor = MapAccessor()

    @JvmStatic
    fun getValue(expression: String, root: Any?): Any? {
        if (root == null) {
            return null
        }
        val context = StandardEvaluationContext()
        context.setRootObject(root)
        if (root is Map<*, *>) {
            context.addPropertyAccessor(mapAccessor)
        }
        return expressionParser.parseExpression(expression).getValue(context)
    }

    @JvmStatic
    fun <T> getValue(expression: String, root: Any?, resultClass: Class<T>): T? {
        if (root == null) {
            return null
        }
        val context = StandardEvaluationContext()
        context.setRootObject(root)
        if (root is Map<*, *>) {
            context.addPropertyAccessor(mapAccessor)
        }
        return expressionParser.parseExpression(expression).getValue(context, resultClass)
    }


}